### PR TITLE
Change `author/about/version` behavior

### LINKED ISSUE #217: Stop doing automatic author, about and version
Since the begining, structopt use the CARGO_PKG environment variables to populate version, about and authors if not provided. As sometime you don't want that to be setted, there is an ugly hack that, if you do `author = ""`, the magic is removed.

That's not so in par with the clap philosophy, as these setters are not called by default on raw clap.

Thus I propose to remove the ugly hack, don't use CARGO_PKG by default, but add without argument options to set these fiels using the CARGO_PKG, i.e.

```rust
#[derive(StructOpt)]
struct Opt {}
```
became
```rust
#[derive(StructOpt)]
#[structopt(about, author, version)]
struct Opt {}
```
and
```rust
#[derive(StructOpt)]
#[structopt(about = "", author = "", version = "")]
struct Opt {}
```
became
```rust
#[derive(StructOpt)]
struct Opt {}
```

To anyone reading this, what do you think?

Would fix #172 and maybe #173 

cc @sphynx @killercup @kbknapp @ssokolow @robinst @0ndorio

### PR DESCRIPTION
**Pull request by CreepySkeleton**
Closes #217

### CODE REVIEW

**Comment by CreepySkeleton** [comment]
I decided not to create a new PR for the `raw` check, I affects the same lines that the first commit does, I'd have to stack my work on top of it anyway

**Comment by CreepySkeleton** [comment]
Interesting...

**Comment by CreepySkeleton** [comment]
@TeXitoi Could you please take a look? It seems like travis crashes on *all* channels with some compiler-bug error. I can't reproduce it locally, my `cargo test` is flawless and shining

**Comment by CreepySkeleton** [comment]
[My travis build](https://travis-ci.org/CreepySkeleton/structopt/builds/575687606) can't reproduce it. Any idea?

**Comment by TeXitoi** [comment]
Strange...

**Comment by CreepySkeleton** [comment]
Yahoo! Now travis is all green just like a troll from 4chan. Can I ask you to download and share build artifacts of [this particular build](https://travis-ci.org/TeXitoi/structopt/builds/575667984?utm_source=github_status&utm_medium=notification) with me? I'd like to dig into this bug or at least report it to rustc team

**Comment by TeXitoi** [comment]
I can't find anything to download. Any pointer?

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,17 +3,30 @@
 ## Breaking changes
 
 * Bump minimum rust version to 1.34 by [@TeXitoi](https://github.com/TeXitoi)
+
+* Update `syn`, `proc_macro2` and `quote` dependencies to version `1`.
```
Comment (TeXitoi) [inline CHANGELOG.md]: That's for the other PR, no?

File: structopt-derive/src/attrs.rs [hunk 2]
```diff
@@ -536,16 +560,60 @@ impl Attrs {
...
+    }
+
+    /// generate methods on top of a field
+    pub fn field_methods(self) -> TokenStream {
+        let methods: Vec<_> = self
             .methods
             .iter()
-            .map(|&Method { ref name, ref args }| quote!( .#name(#args) ));
+            .map(|&Method { ref name, ref args }| quote!( .#name(#args) ))
+            .collect();
```
Comment (TeXitoi) [inline structopt-derive/src/attrs.rs]: I think you don't need to collect, quote take an IntoIterator.

File: structopt-derive/src/attrs.rs [hunk 3]
```diff
@@ -536,16 +560,60 @@ impl Attrs {
...
+
+        self.about
+            .clone()
+            .map(|(ident, arg)| self.push_str_method(ident.into(), arg.into()));
+
+        let methods: Vec<_> = self
+            .methods
+            .iter()
+            .map(|&Method { ref name, ref args }| quote!( .#name(#args) ))
+            .collect();
```
Comment (TeXitoi) [inline structopt-derive/src/attrs.rs]: Same here

File: structopt-derive/src/lib.rs [hunk 4]
```diff
@@ -282,7 +283,7 @@ fn gen_clap(attrs: &[Attribute]) -> GenOutput {
     let attrs = Attrs::from_struct(attrs, Sp::call_site(name), Sp::call_site(DEFAULT_CASING));
     let tokens = {
         let name = attrs.cased_name();
-        let methods = attrs.methods();
+        let methods = attrs.clone().top_level_methods();
```
Comment (TeXitoi) [inline structopt-derive/src/lib.rs]: Why are these clone needed now?

**Comment by CreepySkeleton** [comment]
I don't know, I can't find anything that 

> I can't find anything to download. Any pointer?

I don't know, I can't find any way to do that in google so [I asked](https://stackoverflow.com/questions/57621915/is-there-a-way-to-download-compilation-artifacts-from-a-travis-ci-build) the overmind, let's see what we can do

File: CHANGELOG.md (hunk 1 above)
Comment (CreepySkeleton) [reply CHANGELOG.md]: True, I'll get rid of it

File: structopt-derive/src/attrs.rs (hunk 2 above)
Comment (CreepySkeleton) [reply structopt-derive/src/attrs.rs]: That is correct, the `.collect` is going to be removed

File: structopt-derive/src/attrs.rs (hunk 3 above)
Comment (CreepySkeleton) [reply structopt-derive/src/attrs.rs]: Yes, I'll make it perish

File: structopt-derive/src/lib.rs (hunk 4 above)
Comment (CreepySkeleton) [reply structopt-derive/src/lib.rs]: i just realized that this is quite complicated question, let me think...

**Comment by hellow554** [comment]
> 
> 
> [My travis build](https://travis-ci.org/CreepySkeleton/structopt/builds/575687606) can't reproduce it. Any idea?

Looks like an incremental bug. I think it is covered in https://github.com/rust-lang/rust/issues/63161, you can leave a comment there.

**Comment by CreepySkeleton** [comment]
> 
> 
> > [My travis build](https://travis-ci.org/CreepySkeleton/structopt/builds/575687606) can't reproduce it. Any idea?
> 
> Looks like an incremental bug. I think it is covered in [rust-lang/rust#63161](https://github.com/rust-lang/rust/issues/63161), you can leave a comment there.

Hmm, I like to leave comments, *especially* on my second vacation day :-)

File: structopt-derive/src/lib.rs (hunk 4 above)
Comment (CreepySkeleton) [reply structopt-derive/src/lib.rs]: In my previous attempt these methods mutated `self`. Hence we happen to reuse `attrs` here first call would add methods that are acceptable for `clap::App` but not for `clap::Arg`. First we used this to generate `App` and later we generated `Arg`. My current approach is different, these methods mutate self no longer.

**Comment by CreepySkeleton** [comment]
Also, I squashed all the commits into one, tell me if you want to separate them

**Comment by TeXitoi** [comment]
I squash at merge time. That's simpler if I can review the modifications added, thanks.

**Comment by CreepySkeleton** [comment]
May I ask why do you always squash?

**Comment by TeXitoi** [comment]
One commit, one feature. And some contributors do a lot of commit as "fix typo", "fmt", "oups, bug fix" that can be annoying in a git bisect.

**Comment by CreepySkeleton** [comment]
Yeah, I do understand why people squash meaningless commits like `oops`, `penis` and stuff - but I also find useful to separate related but different things of one giant commit into separate ones. For example, if it was me in your shoes I'd probably break this commit into three - `Change behavior...`, `Add \`raw\` removal warning` and `Update Changelog`. Sometimes feature implementing may grow into one pretty monstrous commit and it's easier to shatter not-really-related things into smaller peaces. IMHO of course...

**Comment by TeXitoi** [comment]
If you do that, I'll just rebase, not squash

**Comment by TeXitoi** [comment]
I don't want to put pressure on contributors about managing commits in a certain way, but I might ask for a separate PR for unrelated stuff. I allowed you to do one PR for raw and change behavior, thus I will not ask you to do 2 PR, but if you separate this work in 2 commit, I'll happily
keep the history with these 2 commits. I will not take the time to separate these 2 commits myself.

**Comment by CreepySkeleton** [comment]
> but if you separate this work in 2 commit, I'll happily
> keep the history with these 2 commits. I will not take the time to separate these 2 commits myself.

Exactly what I would do, except I'd probably just isolate these 2 commits myself. 

N.B: I'm not actually arguing, I just noticed some policy that is different from mine and got curious about it's advantages and drawbacks. That's how people learn, and I'm trying.

**Comment by TeXitoi** [comment]
Thanks!

**Comment by CreepySkeleton** [comment]
Mm, @TeXitoi did you look at travis output? I'm asking because it seems we missed a warning, I just was about to push the fix when you merged it :)

**Comment by TeXitoi** [comment]
Sorry no, just looked at the status

**Comment by sourcefrog** [comment]
It seems like the docs are out of date with this? They seem to still indicate, in https://docs.rs/structopt/0.3.14/structopt/#magical-methods that these default to the Cargo values.

**Pull request merged**